heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type * as TWebpack from 'webpack';
+import type TWebpackDevServer from 'webpack-dev-server';
+import { AsyncParallelHook, AsyncSeriesBailHook, AsyncSeriesHook } from 'tapable';
+import { LegacyAdapters } from '@rushstack/node-core-library';
+import type {
+  HeftConfiguration,
+  IHeftTaskSession,
+  IHeftTaskCleanHookOptions,
+  IHeftTaskPlugin,
+  IHeftTaskRunHookOptions,
+  IScopedLogger
+} from '@rushstack/heft';
+
+import type {
+  IWebpackConfiguration,
+  IWebpackConfigurationWithDevServer,
+  IWebpack4PluginAccessor
+} from './shared';
+import { WebpackConfigurationLoader } from './WebpackConfigurationLoader';
+
+export interface IWebpack4PluginOptions {
+  devConfigurationPath: string | undefined;
+  configurationPath: string | undefined;
+}
+
+/**
+ * @public
+ */
+export const PLUGIN_NAME: 'Webpack4Plugin' = 'Webpack4Plugin';
+const WEBPACK_DEV_SERVER_PACKAGE_NAME: 'webpack-dev-server' = 'webpack-dev-server';
+const WEBPACK_DEV_SERVER_ENV_VAR_NAME: 'WEBPACK_DEV_SERVER' = 'WEBPACK_DEV_SERVER';
+const UNINITIALIZED: 'UNINITIALIZED' = 'UNINITIALIZED';
+
+/**
+ * @internal
+ */
+export default class Webpack4Plugin implements IHeftTaskPlugin<IWebpack4PluginOptions> {
+  private _webpack: typeof TWebpack | undefined;
+  private _webpackConfiguration: IWebpackConfiguration | undefined | typeof UNINITIALIZED = UNINITIALIZED;
+
+  public readonly accessor: IWebpack4PluginAccessor = {
+    hooks: {
+      onLoadConfiguration: new AsyncSeriesBailHook(),
+      onConfigure: new AsyncSeriesHook(['webpackConfiguration']),
+      onAfterConfigure: new AsyncParallelHook(['webpackConfiguration']),
+      onEmitStats: new AsyncParallelHook(['webpackStats'])
+    }
+  };
+
+  public apply(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IWebpack4PluginOptions
+  ): void {
+    // These get set in the run hook and used in the onConfigureWebpackHook
+    let production: boolean;
+    let serveMode: boolean;
+    let watchMode: boolean;
+
+    // Provide the initial webpack configuration by tapping the hook. Tap with stage as
+    // MAX_SAFE_INTEGER to ensure that we run last.
```

Comment:
If we aren't going to allow the hook to be tapped after the built-in operation, why implement it by tapping the hook at all? Why not just have a step that, if there is not a configuration after the accessor hook, does the default load? If the default is on the hook I'd rather have it at a relatively normal stage, then plugins can specify `before: "Webpack4Plugin"`.